internal/controller/apiextensions/usage/reconciler.go

```diff
@@ -0,0 +1,405 @@
+	reasonAddInUseLabel    event.Reason = "AddInUseLabel"
+	reasonRemoveInUseLabel event.Reason = "RemoveInUseLabel"
+	reasonAddFinalizer     event.Reason = "AddFinalizer"
+	reasonRemoveFinalizer  event.Reason = "RemoveFinalizer"
+
+	reasonUsageConfigured event.Reason = "UsageConfigured"
+	reasonWaitUsing       event.Reason = "WaitingUsingDeleted"
+)
+
+type selectorResolver interface {
+	resolveSelectors(ctx context.Context, u *v1alpha1.Usage) error
+}
+
+// Setup adds a controller that reconciles Usages by
+// defining a composite resource and starting a controller to reconcile it.
+func Setup(mgr ctrl.Manager, o apiextensionscontroller.Options) error {
+	name := "usage/" + strings.ToLower(v1alpha1.UsageGroupKind)
+	r := NewReconciler(mgr,
+		WithLogger(o.Logger.WithValues("controller", name)),
+		WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name))))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha1.Usage{}).
+		WithOptions(o.ForControllerRuntime()).
+		Complete(ratelimiter.NewReconciler(name, r, o.GlobalRateLimiter))
+}
+
+// ReconcilerOption is used to configure the Reconciler.
+type ReconcilerOption func(*Reconciler)
+
+// WithLogger specifies how the Reconciler should log messages.
+func WithLogger(log logging.Logger) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.log = log
+	}
+}
+
+// WithRecorder specifies how the Reconciler should record Kubernetes events.
+func WithRecorder(er event.Recorder) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.record = er
+	}
+}
+
+// WithClientApplicator specifies how the Reconciler should interact with the
+// Kubernetes API.
+func WithClientApplicator(c xpresource.ClientApplicator) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.client = c
+	}
+}
+
+// WithFinalizer specifies how the Reconciler should add and remove
+// finalizers to and from the managed resource.
+func WithFinalizer(f xpresource.Finalizer) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.usage.Finalizer = f
+	}
+}
+
+// WithSelectorResolver specifies how the Reconciler should resolve any
+// resource references it encounters while reconciling Usages.
+func WithSelectorResolver(sr selectorResolver) ReconcilerOption {
+	return func(r *Reconciler) {
+		r.usage.selectorResolver = sr
+	}
+}
+
+type usageResource struct {
+	xpresource.Finalizer
+	selectorResolver
+}
+
+// NewReconciler returns a Reconciler of Usages.
+func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
+	kube := unstructured.NewClient(mgr.GetClient())
+
+	r := &Reconciler{
+		client: xpresource.ClientApplicator{
+			Client:     kube,
+			Applicator: xpresource.NewAPIUpdatingApplicator(kube),
+		},
+
+		usage: usageResource{
+			Finalizer:        xpresource.NewAPIFinalizer(kube, finalizer),
+			selectorResolver: newAPISelectorResolver(kube),
+		},
+
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+
+		pollInterval: 30 * time.Second,
```

Comment:
not used anywhere, but it should be.

Reply:
should this be using the global `pollInterval` option from the command line?  If I remember right I think it is applied to all of the internal resource controllers.

Reply:
Fixed by using the global value from the command line and requeing with this value after a successful reconcile.